pkg/services/ngalert/migration/models.go

```diff
@@ -22,69 +21,77 @@ type OrgMigration struct {
-	titleDeduplicatorForFolder func(folderUID string) *migmodels.Deduplicator
-	channelCache               *ChannelCache
-
-	// Migrated folder for a dashboard based on permissions. Parent Folder ID -> unique dashboard permission -> custom folder.
-	permissionsMap        map[int64]map[permissionHash]*folder.Folder
-	folderCache           map[int64]*folder.Folder                      // Folder ID -> Folder.
-	folderPermissionCache map[string][]accesscontrol.ResourcePermission // Folder UID -> Folder Permissions.
-	generalAlertingFolder *folder.Folder
-
-	state *migmodels.OrgMigrationState
+	orgID       int64
+	failOnError bool
+	silences    []*pb.MeshSilence
 }
 
 // newOrgMigration creates a new OrgMigration for the given orgID.
 func (ms *migrationService) newOrgMigration(orgID int64) *OrgMigration {
-	titlededuplicatorPerFolder := make(map[string]*migmodels.Deduplicator)
 	return &OrgMigration{
 		cfg: ms.cfg,
 		log: ms.log.New("orgID", orgID),
 
 		migrationStore:    ms.migrationStore,
 		encryptionService: ms.encryptionService,
 
-		orgID:    orgID,
-		silences: make([]*pb.MeshSilence, 0),
-		titleDeduplicatorForFolder: func(folderUID string) *migmodels.Deduplicator {
-			if _, ok := titlededuplicatorPerFolder[folderUID]; !ok {
-				titlededuplicatorPerFolder[folderUID] = migmodels.NewDeduplicator(ms.migrationStore.CaseInsensitive(), store.AlertDefinitionMaxTitleLength)
-			}
-			return titlededuplicatorPerFolder[folderUID]
-		},
-		channelCache: &ChannelCache{cache: make(map[any]*legacymodels.AlertNotification)},
-
-		permissionsMap:        make(map[int64]map[permissionHash]*folder.Folder),
-		folderCache:           make(map[int64]*folder.Folder),
-		folderPermissionCache: make(map[string][]accesscontrol.ResourcePermission),
-
-		state: &migmodels.OrgMigrationState{
-			OrgID:          orgID,
-			CreatedFolders: make([]string, 0),
-		},
+		orgID: orgID,
+		// failOnError is true if an org migration should fail as soon as an error occurs.
+		// This is mainly for use with the automatic on-startup migration.
+		failOnError: false,
+		silences:    make([]*pb.MeshSilence, 0),
 	}
 }
 
 // ChannelCache caches channels by ID and UID.
 type ChannelCache struct {
-	cache map[any]*legacymodels.AlertNotification
+	channels []*legacymodels.AlertNotification
+	cache    map[any]*legacymodels.AlertNotification
+	fetch    func(ctx context.Context, key notificationKey) (*legacymodels.AlertNotification, error)
 }
 
-func (c *ChannelCache) LoadChannels(channels []*legacymodels.AlertNotification) {
-	for _, channel := range channels {
-		c.cache[channel.ID] = channel
-		c.cache[channel.UID] = channel
+func (c *ChannelCache) Get(ctx context.Context, key notificationKey) (*legacymodels.AlertNotification, error) {
+	if key.ID > 0 {
+		if channel, ok := c.cache[key.ID]; ok {
+			return channel, nil
+		}
+	}
+	if key.UID != "" {
+		if channel, ok := c.cache[key.UID]; ok {
+			return channel, nil
+		}
+	}
+
+	channel, err := c.fetch(ctx, key)
```

Comment:
should we fetch when `UID = "" and ID == 0`?

Reply:
Legacy notifications without an ID or a UID should not be possible, but you're right this is a regression. To keep parity we should skip these malformed entries instead of throw an error. I'll make an explicit check in `extractChannels`